File: web/src/data/dtos/recipe/RecipeDto.ts
import { RecipeIngredientDto } from './RecipeIngredientDto';
import { RecipeStepDto } from './RecipeStepDto';

type BackendRecipePhoto = { url: string } | null;

export class RecipeBaseDto {
    id: number = 0;
    name: string = '';
    tagline: string | null = null;
    category: string | null = null;
    region: string | null = null;
    timeMinutes: number | null = null;
    difficulty: number | null = null;
    status: string = 'published';
    needsPrep: boolean = false;
    dietTags: string[] = [];
    practicalTags: string[] = [];
    photoUrl: string | null = null;

    constructor(obj: Partial<RecipeBaseDto> & { photo?: BackendRecipePhoto }) {
        Object.assign(this, obj);
        if ('photo' in obj) {
            this.photoUrl = obj.photo?.url ?? null;
        }
    }
}

export class RecipeDto extends RecipeBaseDto {
    servings: number = 1;
    description: string | null = null;
    kcalPerServing: number | null = null;

    steps: RecipeStepDto[] = [];
    ingredients: RecipeIngredientDto[] = [];

    constructor(obj: Partial<RecipeDto> & { photo?: BackendRecipePhoto }) {
        super(obj);
        Object.assign(this, obj);
        this.steps = (obj.steps ?? []).map((s) => new RecipeStepDto(s));
        this.ingredients = (obj.ingredients ?? []).map((i) => new RecipeIngredientDto(i));
    }
}

export interface IUpsertRecipeRequest {
    name: string;
    servings: number;
    description?: string | null;
    tagline?: string | null;
    category?: string | null;
    region?: string | null;
    timeMinutes?: number | null;
    difficulty?: number | null;
    kcalPerServing?: number | null;
    photoUrl?: string | null;
    status: string;
    needsPrep: boolean;
    dietTags: string[];
    practicalTags: string[];
    steps: { stepNo: number; section?: string | null; text: string }[];
    ingredients: {
        sortOrder: number;
        section?: string | null;
        ingredientId: number;
        amount?: number | null;
        unit?: string | null;
        amountText?: string | null;
        note?: string | null;
    }[];
}

File: dtos/recipe_dto.go
package dtos

type RecipeDto struct {
	ID             uint                  `json:"id"`
	Name           string                `json:"name"`
	Servings       int                   `json:"servings"`
	Description    *string               `json:"description,omitempty"`
	Tagline        *string               `json:"tagline,omitempty"`
	Category       *string               `json:"category,omitempty"`
	Region         *string               `json:"region,omitempty"`
	TimeMinutes    *int                  `json:"timeMinutes,omitempty"`
	Difficulty     *int                  `json:"difficulty,omitempty"`
	KcalPerServing *int                  `json:"kcalPerServing,omitempty"`
	Photo          *RecipePhotoDto       `json:"photo,omitempty"`
	Steps          []RecipeStepDto       `json:"steps"`
	Ingredients    []RecipeIngredientDto `json:"ingredients"`
}

type RecipeBaseDto struct {
	ID          uint            `json:"id"`
	Name        string          `json:"name"`
	Tagline     *string         `json:"tagline,omitempty"`
	Category    *string         `json:"category,omitempty"`
	Region      *string         `json:"region,omitempty"`
	TimeMinutes *int            `json:"timeMinutes,omitempty"`
	Difficulty  *int            `json:"difficulty,omitempty"`
	Photo       *RecipePhotoDto `json:"photo,omitempty"`
}

type UpsertRecipeRequest struct {
	Name           string  `json:"name"`
	Servings       int     `json:"servings"`
	Description    *string `json:"description"`
	Tagline        *string `json:"tagline"`
	Category       *string `json:"category"`
	Region         *string `json:"region"`
	TimeMinutes    *int    `json:"timeMinutes"`
	Difficulty     *int    `json:"difficulty"`
	KcalPerServing *int    `json:"kcalPerServing"`
	PhotoUrl       *string `json:"photoUrl"`

	Steps       []UpsertStepRequest       `json:"steps"`
	Ingredients []UpsertIngredientRequest `json:"ingredients"`
}

type UpsertStepRequest struct {
	StepNo  int     `json:"stepNo"`
	Section *string `json:"section"`
	Text    string  `json:"text"`
}

type UpsertIngredientRequest struct {
	SortOrder    int      `json:"sortOrder"`
	Section      *string  `json:"section"`
	IngredientId uint     `json:"ingredientId"`
	Amount       *float64 `json:"amount"`
	Unit         *string  `json:"unit"`
	AmountText   *string  `json:"amountText"`
	Note         *string  `json:"note"`
}
